### PR TITLE
i#2283: Adds faulty redzones

### PR DESCRIPTION
**Pull request by johnfxgalea**
Introduces a new option that renders redzones as faulty for umbra_32. When an access, a read or write, is made on a redzone, a fault occurs. 

This is useful for detecting cross boundary accesses without the need to perform explicit checks. 

This functionality is similar to faults upon writes to shared blocks.

I'll write tests soon. I want to get comments on its API, and even more so, ensure that this functionality is wanted. I found it useful/required for my projects.

Issue #2283

### CODE REVIEW

**Comment by johnfxgalea** [comment]
Great, thanks for your reply. What do you think about the API? Put simply, the user does not require to do much, just simply setting a flag in opts. However, redzone sizes are automatically overriden to that of a page size to enable the setting of permissions.

Also, are there current tests that check redzone accesses by any chance?

**Comment by derekbruening** [comment]
> Also, are there current tests that check redzone accesses by any chance?

+ @zhaoqin who created the umbra library who can correct me if I'm missing some tests -- AFAIK, I'm afraid there are not many isolated umbra tests.  Adding some would certainly be appreciated.

**Comment by johnfxgalea** [comment]
@derekbruening I added a test. Please take a look at your convenience. Thanks!

**Comment by johnfxgalea** [comment]
Appveyor again failed due to #2178.

**Comment by derekbruening** [comment]
> Appveyor again failed due to #2178. It seems that these tests are failing constantly and are not flaky.

Yes, Dr. Memory CI is in bad shape.  I tried to figure #2178 out via diagnostics on Appveyor, and tried to repro on various Windows machines, but failed at every step.  I was hoping to upgrade DrM to VS2017 and thus Appveyor to a more recent Windows to "solve" the problem (well, hide the problem and focus on more modern Windows: not enough resources to solve problems on old OS we can't repro).

File: tests/framework/CMakeLists.txt [hunk 1]
```diff
@@ -180,6 +180,15 @@ use_DynamoRIO_extension(umbra_test_insert_app_to_shadow.client drutil)
 target_include_directories(umbra_test_insert_app_to_shadow.client PRIVATE
   ${CMAKE_CURRENT_SOURCE_DIR})
 
+if (NOT X64)
+add_drmf_test(umbra_client_faulty_redzone umbra_app
```
Comment (derekbruening) [inline tests/framework/CMakeLists.txt:184]: indentation missing

File: tests/framework/umbra_client_faulty_redzone.c [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+static bool event_exception_instrumentation(void *drcontext,
+                                            dr_exception_t *excpt);
+#else
+static dr_signal_action_t event_signal_instrumentation(void *drcontext,
+                                                       dr_siginfo_t *info);
+#endif
+
+static void exit_event(void);
+
+DR_EXPORT void dr_client_main(client_id_t id, int argc, const char *argv[]) {
```
Comment (derekbruening) [inline tests/framework/umbra_client_faulty_redzone.c:65]: style: 2 missing newlines

File: tests/framework/umbra_client_faulty_redzone.c [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+  drreg_options_t ops = {sizeof(ops), 4, true};
```
Comment (derekbruening) [inline tests/framework/umbra_client_faulty_redzone.c:66]: style: indent 4 in entire file

File: tests/framework/umbra_client_faulty_redzone.c [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+  if (umbra_insert_app_to_shadow(drcontext, umbra_map, ilist, where, regaddr,
+                                 &scratch, 1) != DRMF_SUCCESS) {
+    DR_ASSERT(false);
+  }
+  instrlist_meta_preinsert(
+      ilist, where,
+      INSTR_XL8(
+          XINST_CREATE_store_1byte(
+              drcontext,
+              OPND_CREATE_MEM8(regaddr, dr_page_size() /* enter redzone */),
```
Comment (derekbruening) [inline tests/framework/umbra_client_faulty_redzone.c:135]: So this is deliberately *not* accessing the proper shadow xl8, but instead jumping a page ahead, for the sole purpose of testing the redzone, right?  Maybe worth a longer comment.

File: umbra/umbra.h [hunk 5]
```diff
@@ -245,6 +245,22 @@ typedef struct _umbra_map_options_t {
     ptr_uint_t redzone_value;
     /** The redzone value size, only 1 is supported now */
     size_t     redzone_value_size;
+    /**
+     * Set to true to render redzones as faulty, i.e., access (both read and write)
+     * to a redzone causes a fault.
+     *
+     * If a cross block access occurs, a fault is triggered. This is an
+     * optimisation, because rather than requiring the user to perform explicit
+     * checks on redzone access, a fault will indicate the case. User may
```
Comment (derekbruening) [inline umbra/umbra.h:254]: nit: The user

File: umbra/umbra.h [hunk 6]
```diff
@@ -245,6 +245,22 @@ typedef struct _umbra_map_options_t {
...
+     * then take further action by defining a fault handler, similar to
+     * that done for shared block access (if enabled).
+     *
+     * Overrides redzone data specified in the struct, including redzone_size.
+     * With this option enabled, the fields: redzone_size, redzone_value and
+     * redzone_value_size, are not considered. Redzone size is set to a page size
```
Comment (derekbruening) [inline umbra/umbra.h:260]: Add `#` prefixes for doxygen links to the fields.

File: umbra/umbra_32.c [hunk 7]
```diff
@@ -104,7 +133,19 @@ static byte *
...
-    block = global_alloc(map->shadow_block_alloc_size, HEAPSTAT_SHADOW);
+
+    if (map->options.make_redzone_faulty) {
+        /* We need to ensure page alignment for redzones, so that we may then
+         * assign them no access.
+         */
+        block = nonheap_alloc(map->shadow_block_alloc_size,
+                              DR_MEMPROT_READ | DR_MEMPROT_WRITE,
+                              HEAPSTAT_SHADOW);
+        } else {
```
Comment (derekbruening) [inline umbra/umbra_32.c:144]: indentation is off

File: umbra/umbra_32.c [hunk 8]
```diff
@@ -531,6 +572,19 @@ umbra_arch_exit()
 drmf_status_t
 umbra_map_arch_init(umbra_map_t *map, umbra_map_options_t *ops)
 {
+    /* Override redzone size. We must make it a page size to control
+     * its permissions.
+     *
+     * FIXME i#2283: Potentially support larger redzone sizes to be even more safe.
```
Comment (derekbruening) [inline umbra/umbra_32.c:578]: nit: s/FIXME/XXX/
(not a correctness issue)

File: tests/framework/CMakeLists.txt (hunk 1 above)
Comment (johnfxgalea) [reply tests/framework/CMakeLists.txt:184]: Done.

File: tests/framework/umbra_client_faulty_redzone.c (hunk 2 above)
Comment (johnfxgalea) [reply tests/framework/umbra_client_faulty_redzone.c:65]: Done.

File: tests/framework/umbra_client_faulty_redzone.c (hunk 3 above)
Comment (johnfxgalea) [reply tests/framework/umbra_client_faulty_redzone.c:66]: Done.

File: tests/framework/umbra_client_faulty_redzone.c (hunk 4 above)
Comment (johnfxgalea) [reply tests/framework/umbra_client_faulty_redzone.c:135]: Done.

File: umbra/umbra.h (hunk 5 above)
Comment (johnfxgalea) [reply umbra/umbra.h:254]: Done.

File: umbra/umbra.h (hunk 6 above)
Comment (johnfxgalea) [reply umbra/umbra.h:260]: Done.

File: umbra/umbra_32.c (hunk 7 above)
Comment (johnfxgalea) [reply umbra/umbra_32.c:144]: Done.

File: umbra/umbra_32.c (hunk 8 above)
Comment (johnfxgalea) [reply umbra/umbra_32.c:578]: Done.

**Comment by johnfxgalea** [comment]
@derekbruening PTAL

File: umbra/umbra.h [hunk 9]
```diff
@@ -50,18 +50,23 @@ extern "C" {
...
-# define dr_insert_write_tls_field DO_NOT_USE_tls_field_USE_drmgr_tls_field_instead
-# define dr_register_thread_init_event DO_NOT_USE_thread_event_USE_drmgr_events_instead
-# define dr_unregister_thread_init_event DO_NOT_USE_thread_event_USE_drmgr_events_instead
-# define dr_register_thread_exit_event DO_NOT_USE_thread_event_USE_drmgr_events_instead
-# define dr_unregister_thread_exit_event DO_NOT_USE_thread_event_USE_drmgr_events_instead
-# define dr_register_pre_syscall_event DO_NOT_USE_pre_syscall_USE_drmgr_events_instead
-# define dr_unregister_pre_syscall_event DO_NOT_USE_pre_syscall_USE_drmgr_events_instead
-# define dr_register_post_syscall_event DO_NOT_USE_post_syscall_USE_drmgr_events_instead
-# define dr_unregister_post_syscall_event DO_NOT_USE_post_syscall_USE_drmgr_events_instead
+#    define dr_get_tls_field DO_NOT_USE_tls_field_USE_drmgr_tls_field_instead
```
Comment (derekbruening) [inline umbra/umbra.h:53]: Probably best to leave big changes to unrelated code to a separate commit: it will make it hard to see what this PR itself is changing.

Comment (johnfxgalea) [reply umbra/umbra.h:53]: Can I revert back with bad formatting and then, immediately after, do a separate PR with proper formatting on umbra files that is done with clang-format please?

Comment (derekbruening) [reply umbra/umbra.h:53]: I think the indent/format issues in this PR were all in the new test and not in umbra.h or umbra_32.c?  Or if there were any in those files they were just one or two which could be manually fixed.  Seems like this PR can go in w/o adding bad formatting (or at least that I noticed in my review)?  Afterward, sure, if you want to clang-format the umbra files and set up a CI check: that's part of #2107.

**Comment by johnfxgalea** [comment]
@derekbruening PTAL

**Comment by johnfxgalea** [comment]
@derekbruening Thank you for the review!

**Comment by johnfxgalea** [comment]
@derekbruening Can you point me where I can find how Dynamorio does the format check please?

**Comment by derekbruening** [comment]
> @derekbruening Can you point me where I can find how Dynamorio does the format check please?

I put a pointer in https://github.com/DynamoRIO/drmemory/issues/2107#issuecomment-651174150
